src/components/Highlight/Highlight.vue

```diff
@@ -0,0 +1,184 @@
+  -
+  - @author Raimund Schlüßler <raimund.schluessler@mailbox.org>
+  -
+  - @license GNU AGPL version 3 or any later version
+  -
+  - This program is free software: you can redistribute it and/or modify
+  - it under the terms of the GNU Affero General Public License as
+  - published by the Free Software Foundation, either version 3 of the
+  - License, or (at your option) any later version.
+  -
+  - This program is distributed in the hope that it will be useful,
+  - but WITHOUT ANY WARRANTY; without even the implied warranty of
+  - MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  - GNU Affero General Public License for more details.
+  -
+  - You should have received a copy of the GNU Affero General Public License
+  - along with this program. If not, see <http://www.gnu.org/licenses/>.
+  -
+  -->
+
+<docs>
+
+### General description
+
+Highlight a string with html &lt;strong&gt;. Accepts a substring to highlight or an array with ranges.
+
+### Usage
+
+```vue
+<template>
+	<div>
+		<Highlight text="Highlight me please!" search="me" />
+		<br />
+		<Highlight text="Highlight me please!" :highlight="[{ start: 4, end: 12 }]" />
+	</div>
+</template>
+```
+</docs>
+
+<script>
+export default {
+	name: 'Highlight',
+	props: {
+		/**
+		 * The string to display
+		 */
+		text: {
+			type: String,
+			default: '',
+		},
+		/**
+		 * The string to match and highlight
+		 */
+		search: {
+			type: String,
+			default: '',
+		},
+		/**
+		 * The ranges to highlight, takes precedence over the search prop.
+		 */
+		highlight: {
+			type: Array,
+			default: () => [],
+		},
+	},
+	computed: {
+		/**
+		 * The indice ranges which should be highlighted.
+		 * If an array with ranges is provided, we use it. Otherwise
+		 * we calculate it based on the provided substring to highlight.
+		 *
+		 * @returns {Array} The array of ranges to highlight
+		 */
+		ranges() {
+			let ranges = []
+			// If the search term and the highlight array is empty, return early with empty array
+			if (!this.search && this.highlight.length === 0) return ranges
+
+			// If there are ranges to highlight provided, we use this array.
+			if (this.highlight.length > 0) {
+				ranges = this.highlight
+			// Otherwise we check the text to highlight for matches of the search term.
+			} else {
+				let currentIndex = 0
+				let index = this.text.toLowerCase().indexOf(this.search.toLowerCase(), currentIndex)
+				while (index > -1) {
+					currentIndex = index + this.search.length
+					ranges.push({ start: index, end: currentIndex })
+
+					index = this.text.toLowerCase().indexOf(this.search.toLowerCase(), index + 1)
+				}
+			}
+
+			/**
+			 * Validate the ranges array to be within the string length
+			 * and discard ranges which are completely out of bonds.
+			 */
+			return ranges.reduce((ranges, range) => {
+				if (range.start < this.text.length && range.end > 0) {
+					ranges.push({
```

Comment:
I see two red flags here.

1) The array is modified as you iterate over it.
2) You modify the array with `.push` while you could just return a new array.

Reply:
The two `ranges` are different arrays. The `ranges` in the callback (`(ranges, range) => {}`) is the accumulator of the `reduce` function, which is initialized with `[]`, pushed to in the callback and then returned as the result of the reduce function. https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/Reduce

But I can change the names of the variables to make that more clear.